lib/cylc/network/server.py

```diff
@@ -278,7 +286,7 @@ def __init__(self, schd):
             self,
             encrypt,
             decrypt,
-            lambda: get_secret(schd.suite)
+            partial(get_secret, schd.suite)
```

Comment:
**TLDR;** lambdas are considered by some (incl guido) to have been a mistake, one of "Python's glitches"!

There are *some* benefits of `partial` including:
* There is apparently a slight improvement in performance.
* `functools.partial` is also a little more Pythonic than the a with Python as [it understands that it is an adapter and knows what it is passing arguments through to](https://stackoverflow.com/questions/24532532/functools-partial-vs-normal-python-function/24532589#24532589) which presumably means it can work nicely with IDEs etc.

But really it's about lambas falling from favour rather than partial delivering superiour functionality, nowerdays Python has turned against lambdas:
* They were never properly implemented (e.g. can only be defined on one line)
* Named lambdas break `pycodestyle` (the argument being you should just use a function instead)
* Late binding issues cause endless confusion ([for example](https://stackoverflow.com/questions/1107210/python-lambda-problems)).

https://stackoverflow.com/a/3252425

Reply:
Interesting! I've never used `functools.partial` 🤔

Reply:
Gotcha! But in this case lambdas and partials deliver pretty much the same feature, right?

In Java/Scala lambdas and closures are still being used, and the equivalent for partials are currying in Scala (supported out of the box by the language by simply not providing all parameters) and in Java by (as verbose as always) creating some objects instantiating classes from `java.util.function`.

I think both are still used, in some cases developers prefer lambdas for clarity/simplicity, while others prefer currying to be closer for FP. And there are cases when one or the other can be misused causing unwanted bugs, performance regressions, etc. But I think none are discouraged yet.

If I understood correctly, in this case, we have something similar to this code:

```
from functools import partial


class Scheduler(object):

    def __init__(self):
        self.suite = "sugar"
        self.testa = lambda: self.echovalue(self.suite)
        self.testb = partial(self.echovalue, self.suite)

    def echovalue(self, value):
        print(value)


s = Scheduler()
s.testa()  # sugar
s.testb()  # sugar

s.suite = "salt"
s.testa()  # salt
s.testb()  # sugar, because partial is similar to a... FP applicative functor? keeping the value inside it.. I think...
```

So I think it means that if we ever touch `sched.suite` and change its value, then the partial object we passed would be outdated, and we would actually have to create a new one. I think?

So much to learn in Python :worried: and quite late here, so sorry if wrote anything silly :coffee:

Reply:
> So much to learn in Python worried and quite late here, so sorry if wrote anything silly coffee

Nothing silly whatsoever.

> Gotcha! But in this case lambdas and partials deliver pretty much the same feature, right?

Exactly!

> In Java/Scala lambdas and closures are still being used

Lambdas can be nice, but as a concept they have never really been properly integrated into Python. Their scope is limited, only one expression and no line breaks permitted.

> But I think none are discouraged yet.

Lambdas are discouraged in Python to some extent if not entirely:

* List comprehensions and generator expressions are preferred to lambdas where possible removing most lambda use cases.
* Named lambdas are actively discouraged.
* The reason that the `map` function is discouraged is actually to do with the `lambda` use case e.g. `map([1, 2, 3], lambda x: x**2)`.

> So I think it means that if we ever touch sched.suite and change its value, then the partial object we passed would be outdated

Yep, to me that's the correct behaviour!

Late closures can be useful but I think they are likely to catch you out more often than they are to come in handy.

Reply:
>Yep, to me that's the correct behaviour!

Well, I'm -0 here. Fine if others (cc @cylc/core) prefer the `partial`. For me if I need to change the `.suite` property of `Scheduler`, I wouldn't want to have to remember that I also need to update the encode and the decode method.

>Late closures can be useful but I think they are likely to catch you out more often than they are to come in handy.

Agreed, but in this case I think I would stick with the lambda to prevent a possible bug. -0 and -1 because we don't change the value of `.suite` now. As far as I can tell, renaming a suite causes that suite to be re-registered, creating a new instance of `Scheduler`. But if that changed later to actually reuse the `Scheduler` (for reasons), then we would have a bug where the suite was not able to authenticate I think.